### PR TITLE
Issue634 upload form

### LINKED ISSUE #634: Upload Form Mega Project
- [x] generate from existing simulation
 - [x] expand number of x y z fields in data section based on 2D versus 3D button
 - [x] add `type` field to meta.yaml from upload form
 - [x] ensure `counter` data list works when adding and deleting data in the wrong order
 - [x] add media file section
 - [x] populate data & media files from previous section
 - [x] ensure responsiveness of form
 - [x] test and check it works
   - [x] will need to modify schema
     - [x] long title (descriptive title)
     - [x] multiple authors
     - [x] add field descritiption `desc:` in PyKwalify schema fields and use on webform, http://www.kuwata-lab.com/kwalify/ruby/users-guide.01.html#schema, this can also be used for the DOI
     - [x] remove details sections
     - [x] have one software section or implementation 
   - [x] ensure that landing page for each simulation are correct
   - [x] ensure that upload pages are correct for combined results
   - [x] ensure that simulation table renders correctly
   - [x] ensure that upload from is populated correctly for each simulation
   - [x] test modifying a simulation with a new name
   - [x] test updating a simulation
   - [x] test completely new simulations

### PR DESCRIPTION
**Pull request by wd15**
Fix #634

## Changes

 - Improve the upload form by allowing an arbitrary number of data and media files.

 - Enable populating of the form from existing results.

 - Simplify the metadata schema.

## Issues

Staticman does not [let existing data files update](https://github.com/eduardoboucas/staticman/pull/126) so a new simulation name is created even for an update. I might address this in later versions by having a live or dead flag in the meta.yaml files and also decoupling the simulation name from the directory name and having a unique ID for the directory names instead.

 - The tooltips still don't work. That is in a different [issue](https://github.com/usnistgov/chimad-phase-field/issues/385) to be addressed soon.

## How to Review

These changes can be [viewed live][live-site].

 - Check that you like the look and feel of the [upload form](http://random-cat-668.surge.sh/simulations/upload_form/).

 - Try entering data and submitting. Name the submission (short name) "fake" so that I know it is a fake submission

 -  Try populating from an existing simulation. Go to the [simulation result table](http://random-cat-668.surge.sh/simulations/#simulations). Click on a simulation link - it should open up an individual simulation result landing page. Click on the purple edit button to the side of name of the simulation. It should populate the form. Try changing (or not) some of the data in the form and submitting.

 - Try submitting new data files and media files - anything will do, doesn't have to be real data

**Remember to tear down the [live site][live-site] when merging or
  closing this pull-request.**

[live-site]: http://random-cat-668.surge.sh

---
This change is [Reviewable](https://reviewable.io/reviews/usnistgov/chimad-phase-field/668)

### CODE REVIEW

**Comment by tkphd** [comment]
The corner-ribbon has URL=http://random-cat-668.surge.sh/simulations/upload_form (works)

The upload-button has URL=http://random-cat-668.surge.sh/chimad-phase-field/simulations/upload_form (404)

**Comment by tkphd** [comment]
When updating an existing simulation (e.g., changing the parallel model), the staticman PR creates a new folder by appending `_update` to the old one. If the PR is merged as-is, it is a duplication, not an update, and there's no diff to help us check what's been changed.

Can the form actually update an existing result?

**Comment by wd15** [comment]
@tkphd: regarding the `_update` thing. I described the problem under "Issues" in the description of the pull request.

To reiterate, currently, Staticman does not let you overwrite the same file. Hence why `_update` is added to the name of the directory. You can try do it with the same name and see what happens, you'll see that you get an error. As far as I can tell, there is not anyway to override this short of fixing Staticman. There is a [pull request](https://github.com/eduardoboucas/staticman/pull/126) to do this to which I've given a thumbs up.

One sort of [work around](https://github.com/usnistgov/chimad-phase-field/issues/674) is to start using unique IDs for the simulation directories and move the name of the simulation into the metadata (this still won't let you see diffs of course). Also, there is no way to delete simulations from the frontend directly for this reason. 

Basically, it's not something that fixable right now.

**Comment by wd15** [comment]
@tkphd: fixed the broken upload button

**Comment by tkphd** [comment]
Reviewed 78 of 78 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[_includes/simulation.html, line 109 at r1](https://reviewable.io:443/reviews/usnistgov/chimad-phase-field/668#-L2bIzHCjCfPN6kA1G57:-L2bIzHCjCfPN6kA1G58:b-fxwoui) ([raw file](https://github.com/usnistgov/chimad-phase-field/blob/9f1d511bf3116923e8af0690d77918f7b976224f/_includes/simulation.html#L109)):*
> ```HTML
> 
> <script>
>   var SIM_NAME = new URL(window.location.href).searchParams.get("sim");
> ```

I'm not familiar with dynamic memory in Java. Does this require a matching `SIM_NAME=null;`, or is the heap location freed when the script exits?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/usnistgov/chimad-phase-field/668)*

**Comment by wd15** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[_includes/simulation.html, line 109 at r1](https://reviewable.io:443/reviews/usnistgov/chimad-phase-field/668#-L2bIzHCjCfPN6kA1G57:-L2bOvbylrZM0BHlrVhu:b-twdryz) ([raw file](https://github.com/usnistgov/chimad-phase-field/blob/9f1d511bf3116923e8af0690d77918f7b976224f/_includes/simulation.html#L109)):*
<details><summary><i>Previously, tkphd (Trevor Keller) wrote…</i></summary><blockquote>

I'm not familiar with dynamic memory in Java. Does this require a matching `SIM_NAME=null;`, or is the heap location freed when the script exits?
</blockquote></details>

This isn't Java, it's Javascript. Totally different language. Javascript does its own garbage collection and `new` has nothing to do with memory allocation. See https://stackoverflow.com/questions/1646698/what-is-the-new-keyword-in-javascript.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/usnistgov/chimad-phase-field/668)*

**Comment by stvdwtt** [comment]
I'll try to take a look this afternoon and post my review.

**Comment by wd15** [comment]
@stdwtt, thanks for reviewing!

On Wed, Jan 17, 2018 at 2:38 PM, Stephen DeWitt
<notifications@github.com> wrote:
>
> It seems like there's an extra blank line between the "Summary" text and the green line indicating that text has been entered:

This indicates a "textarea" element that includes new lines as opposed
to a regular "input" element that does not allow new lines. The
styling is entirely based on Materialize CSS and I'm loathe to change
it. Is that cool? See http://materializecss.com/forms.html#textarea
for more details. "textarea" enable one to copy and paste a markdown
file for example or a readme describing the code. You can't do that
with regular "input" elements.

> The same thing happens in the "Data Description" boxes.

Same thing, they are "textarea" elements.

> I think "Name of first column" and "Name of second column" would be clearer than "Name of x field" and "Name of y field" in the "Data Files" section. For a line cut along y, the name of the first column would be "y", which is confusing.

I'm indicating the "x field" and "y field" on a 2D plot or 3D contour
plot that the data will be used to display on the website. I'm not
referring to the column in the CSV file. The CSV file could have many
columns and the "x field" doesn't necessarily correspond to the first
column. I could use "x axis" instead of "x field" for example. I've
changed it to "Name of x-axis column".

> I get this error after hitting "Submit"

Site is broken right now as we're moving to pfhub so nothing works.
I'll let you know when it's working so you can retest. See #681.

> It isn't clear when clicking the edit button on an existing simulation if the new input supersedes the existing results. Is the primary use of this to edit an existing result or to prepopulate the form from a previous result to save time inputting? If the goal is to prepopulate the form for a new result, there should be some text on the upload page to that effect. Otherwise I wouldn't know to go to a previous result.

I've updated the text to indicate that you can repopulate from an
existing result. Also, see the earlier discussion with @tkphd about
the subtleties with this issue and using Staticman.

-- 
Daniel Wheeler

**Comment by stvdwtt** [comment]
@wd15 

RE the "textarea" element: That makes sense. I haven't noticed that convention elsewhere, but it makes sense to give a visual distinction between expandable and non-expandable text.

RE the axis labels: Good point about the order/number columns in the csv file not necessarily matching up with what we're requesting. Thanks for making the change to "Name of ?-axis column", that's clearer. 

RE the site being down: I wondered if that was the case.

RE the edit button: Thanks for adding that text. However, regardless of how we/Staticman handle it in the background, it needs to be clear to users whether they're editing a previous simulation or they're using a previous simulation as a template for a new simulation.

**Comment by wd15** [comment]
@stdwtt: currently we can't edit existing simulations due to the Staticman problem. What should I change to make that better in terms of the edit button. Shall I have some hover text over that button explaining that this can be used to derive a new simulation?

**Comment by wd15** [comment]
@stvdwtt: added hover text for the edit button.

**Comment by stvdwtt** [comment]
@wd15 If the intended use of the edit button is to create a new entry similar to the existing entry, I'd recommend replacing the edit button with a rectangular button that has text to the effect of "Use as template for new submission". Using the edit button makes it seem like the intended use is to edit the existing entry.

**Comment by wd15** [comment]
On Thu, Jan 18, 2018 at 3:52 PM, Stephen DeWitt
<notifications@github.com> wrote:
>
> @wd15 If the intended use of the edit button is to create a new entry similar to the existing entry, I'd recommend replacing the edit button with a rectangular button that has text to the effect of "Use as template for new submission". Using the edit button makes it seem like the intended use is to edit the existing entry.

@stvdwtt The button has been changed to have the text suggested.

Also the website is back up so the form works again. Please try
submitting some data at
http://random-cat-668.surge.sh/simulations/upload_form/



-- 
Daniel Wheeler

**Comment by wd15** [comment]
On Fri, Jan 19, 2018 at 8:37 AM, Stephen DeWitt
<notifications@github.com> wrote:
>
> Minor quibble: Is there a reason that i386 is the default CPU architecture? x86-64 is more common isn't it? (Or maybe I'm wrong, processor architecture isn't my forte.) I thought in the past that x86-64 was the default.

Changed now.

> Question: What is the intended difference between "Wall Clock Time" and "Execution End Time"? I took a look at the branch with the tooltips fixed and I'm still not sure.

Wall clock time is the real world time taken to run the simulation
from start to finish. The simulation time (not real world time)
traversed / spanned from the start to finish. What is a better term
than "Execution End Time"? I think I did have "Simulation End Time".
How do you differentiate between simulation time and wall clock time?
I think wall clock time is clear to most people. I think the issue is
with simulation time, but  I actually agree that "Execution End Time"
could refer to either time scale. I'm going to change ti back to
"Simulation End Time" until we think of something better. Anyway, I'm
going to improve the tooltips based on what I've written here in #683
unless you have some more suggestions.

> Comment outside the scope of this PR:
> If possible, I'd suggest staying on the PFHub site instead of going to GitHub after submitting an upload. I think a page saying "Upload Complete!" with a link to the Staticman generated PR would feel more intuitive.

For sure, see #693.

> Thanks for the updates!

Thanks for the review. You deserve the reviewer of the month badge.

-- 
Daniel Wheeler

**Comment by stvdwtt** [comment]
You're very welcome, glad to help.

I think "Simulations end time" was supposed to be "Simulation end time", right? Assuming that's the case, I made the edit. 

I think you're right that "Simulations end time" is clearer than "Execution end time" and with a tooltip should work nicely.

**Comment by stvdwtt** [comment]
Hmm, the change I made to "Simulations end time" isn't showing up on the random-cat test site. Do I have to wait until the Travis check is done for it to go up?

**Comment by wd15** [comment]
On Fri, Jan 19, 2018 at 1:22 PM, Stephen DeWitt
<notifications@github.com> wrote:
>
> You're very welcome, glad to help.
>
> I think "Simulations end time" was supposed to be "Simulation end time", right? Assuming that's the case, I made the edit.

I'm totally confused. How on earth did you make the edit? The branch
is in my local "wd15/pfhub" repository. Do you have access to commit
to that? That's crazy.

-- 
Daniel Wheeler

**Comment by wd15** [comment]
On Fri, Jan 19, 2018 at 1:26 PM, Stephen DeWitt
<notifications@github.com> wrote:
> Hmm, the change I made to "Simulations end time" isn't showing up on the
> random-cat test site. Do I have to wait until the Travis check is done for
> it to go up?

Right, it only gets pushed to Surge at the end of the test. See
https://github.com/usnistgov/pfhub/blob/master/.travis.yml#L39

**Comment by stvdwtt** [comment]
> >On Fri, Jan 19, 2018 at 1:26 PM, Stephen DeWitt ***@***.***> wrote:
 Hmm, the change I made to "Simulations end time" isn't showing up on the
 random-cat test site. Do I have to wait until the Travis check is done for
 it to go up?

>Right, it only gets pushed to Surge at the end of the test. See
https://github.com/usnistgov/pfhub/blob/master/.travis.yml#L39

Ok, that makes sense.

> I'm totally confused. How on earth did you make the edit? The branch
is in my local "wd15/pfhub" repository. Do you have access to commit
to that? That's crazy.

Yeah, I wasn't actually expecting it to work and was surprised when it did.
